bigquery/storage/managedwriter/managed_stream.go

```diff
@@ -140,3 +159,173 @@ func (ms *ManagedStream) Finalize(ctx context.Context) (int64, error) {
+	var r gax.Retryer = &defaultRetryer{}
+	if settings.Retry != nil {
+		r = settings.Retry()
+	}
+
+	var arc *storagepb.BigQueryWrite_AppendRowsClient
+	var ch chan *pendingWrite
+	var err error
+
+	for {
+		arc, ch, err = ms.getStream(arc)
+		if err != nil {
+			return err
+		}
+		err = f(*arc, ch)
+		if err != nil {
+			bo, shouldRetry := r.Retry(err)
+			if shouldRetry {
+				if err := gax.Sleep(ms.ctx, bo); err != nil {
+					return err
+				}
+				continue
+			}
+			ms.mu.Lock()
+			ms.err = err
+			ms.mu.Unlock()
+		}
+		return err
+	}
+}
+
+func (ms *ManagedStream) append(pw *pendingWrite) error {
+	return ms.call(func(arc storagepb.BigQueryWrite_AppendRowsClient, ch chan *pendingWrite) error {
+		// TODO: we should only send stream ID and schema for the first message in a new stream, but
+		// we've decoupled the individual appends from caring about state of the connection.  We likely
+		// need to do this via a mutex-guarded state flag.
+		pw.request.WriteStream = ms.streamSettings.streamID
+		pw.request.GetProtoRows().WriterSchema = &storagepb.ProtoSchema{
+			ProtoDescriptor: ms.schemaDescriptor,
+		}
+		err := arc.Send(pw.request)
+		if err == nil {
+			ch <- pw
+		}
+		return err
+	})
+}
+
+func (ms *ManagedStream) CloseSend() error {
+	err := ms.call(func(arc storagepb.BigQueryWrite_AppendRowsClient, ch chan *pendingWrite) error {
+		err := arc.CloseSend()
+		if err == nil {
+			close(ch)
+		}
+		return err
+	})
+	ms.mu.Lock()
+	ms.err = io.EOF
+	close(ms.pending)
+	ms.mu.Unlock()
+	return err
+}
+
+// AppendRows sends the append requests to the service, and returns one AppendResult per row.
+func (ms *ManagedStream) AppendRows(data [][]byte, offset int64) ([]*AppendResult, error) {
+	pw := newPendingWrite(data, offset)
+	if err := ms.append(pw); err != nil {
+		// pending write is DOA, mark it done.
+		pw.markDone(NoStreamOffset, err)
+		return nil, err
+	}
+	return pw.results, nil
+}
+
+// recvProcessor is used to propagate append responses back up with the originating write requests in a goroutine.
+func recvProcessor(ctx context.Context, arc storagepb.BigQueryWrite_AppendRowsClient, ch <-chan *pendingWrite) {
+	// TODO:  We'd like to re-send requests that are in an ambiguous state due to channel errors.  For now, we simply
+	// ensure that pending writes get acknowledged with a terminal state.
+	for {
+		select {
+		case <-ctx.Done():
+			// Context is done, so we're not going to get further updates.  Mark all work failed with the context error.
+			for {
+				pw, ok := <-ch
+				if !ok {
+					return
+				}
+				pw.markDone(NoStreamOffset, ctx.Err())
+			}
+		case nextWrite, ok := <-ch:
+			if !ok {
+				// Channel closed, all elements processed.
+				return
+			}
+
+			// block until we get a corresponding response or err from stream.
+			resp, err := arc.Recv()
+			if err != nil {
+				log.Printf("recv got err: %#v", err)
+				nextWrite.markDone(NoStreamOffset, err)
+			}
+
+			if status := resp.GetError(); status != nil {
```

Comment:
Is this needed? Is the error not already a grpc.Status in this case?

Reply:
the response is a oneof that can either signal append error of append success.

Reply:
Oh streaming RPC, duh. My bad.